cmd/deploy/deploy.go

```diff
@@ -212,48 +213,42 @@ func (dc *DeployCommand) RunDeploy(ctx context.Context, deployOptions *Options)
 		}
 
 		oktetoLog.Information("Checking to build services at manifest")
-		if deployOptions.Manifest.Build != nil {
-
-			for service, mOptions := range deployOptions.Manifest.Build {
-				buildOptions := build.BuildOptions{OutputMode: oktetoLog.GetOutputFormat()}
-				if mOptions.Name == "" {
-					mOptions.Name = deployOptions.Name
+		if deployOptions.Build {
+			// force build of all services
+			for service, mBuildInfo := range deployOptions.Manifest.Build {
+				if err := runBuildAndSetEnvs(ctx, service, mBuildInfo); err != nil {
+					return err
 				}
-				opts := build.OptsFromManifest(service, mOptions, buildOptions)
+			}
 
-				if deployOptions.Build {
-					oktetoLog.Debug("force build from manifest definition")
-					if err := runBuildAndSetEnvs(ctx, service, opts); err != nil {
-						return err
-					}
-					continue
-				}
+		} else {
+			// check if images are at registry (global or dev) and set envs or send to build
+			toBuild := make(chan string, len(deployOptions.Manifest.Build))
+
+			g, _ := errgroup.WithContext(ctx)
+			for service, mBuildInfo := range deployOptions.Manifest.Build {
+				service, mBuildInfo := service, mBuildInfo
+				g.Go(func() error {
+					return checkServicesToBuild(service, mBuildInfo, toBuild)
+				})
+			}
 
-				if build.ShouldOptimizeBuild(opts.Tag) {
-					oktetoLog.Debug("found OKTETO_GIT_COMMIT, optimizing the build flow")
-					if skipBuild, err := checkImageAtGlobalAndSetEnvs(service, opts); err != nil {
-						return err
-					} else if skipBuild {
-						continue
-					}
-				}
+			if err := g.Wait(); err != nil {
+				return err
+			}
+			close(toBuild)
 
-				if imageWithDigest, err := registry.GetImageTagWithDigest(opts.Tag); err == oktetoErrors.ErrNotFound {
-					oktetoLog.Debugf("image not found, building image %s", opts.Tag)
-					if err := runBuildAndSetEnvs(ctx, service, opts); err != nil {
-						return err
-					}
-					continue
-				} else if err != nil {
-					return fmt.Errorf("error checking image at registry %s: %v", opts.Tag, err)
-				} else {
-					oktetoLog.Success("Skipping build for image for service %s found: %s", service, opts.Tag)
-					if err := setManifestEnvVars(service, imageWithDigest); err != nil {
-						return err
-					}
+			for {
+				svc, ok := <-toBuild
+				if svc == "" || !ok {
```

Comment:
Shouldn't we do the break only if there are no more elements (`!ok`)? Could it happen that some serviceToBuild is sent as empty `svc == ""` but there are still some more elements in the channel? In that case, maybe we should do a continue if `svc == "" && ok`

Reply:
it should break if `!ok` so the `srv == ""` condition is no longer applicable, as this will be only when the channel is closed and is not returning anything. the toBuild channel only returns when the image has to be build so i cant think of anycase where it returns an empty string within the channel being open. thnks!